File: backend/src/main/java/com/vidacotidiana/shared/domain/ModuleContext.java
package com.vidacotidiana.shared.domain;

/**
 * ADR-019: módulo propietario de un recurso — Personal o Laboral.
 *
 * Pedido explícito del usuario (2026-08-28): "cada recurso debe pertenecer
 * exclusivamente al módulo desde el cual fue creado... el aislamiento debe
 * existir también en la lógica de negocio y en las consultas, no únicamente
 * ocultando elementos mediante filtros visuales".
 *
 * Vive en `shared` y no en un módulo concreto porque lo comparten cuatro
 * agregados sin relación entre sí (Garantía, Mantenimiento, Suscripción,
 * Nota del día). El enum `reminder.domain.ReminderContext` se deja
 * intacto: ya existía, ya está persistido y cambiarlo obligaría a tocar el
 * módulo `reminder` entero sin ninguna ganancia — son dos representaciones
 * del mismo concepto que conviven, y `ReminderService` sigue siendo su
 * único dueño.
 *
 * REGLA DE MIGRACIÓN (V25): todo lo que ya existía es PERSONAL. No se
 * reinterpreta nada como LABORAL.
 */
public enum ModuleContext {
    PERSONAL,
    LABORAL;

    /**
     * Resuelve el contexto recibido de la API. Nulo o vacío ⇒ PERSONAL: es
     * el valor que tenían todos los recursos anteriores a esta regla y el
     * que corresponde a cualquier pantalla que aún no envía contexto.
     */
    public static ModuleContext fromNullable(String raw) {
        if (raw == null || raw.isBlank()) {
            return PERSONAL;
        }
        return ModuleContext.valueOf(raw.trim().toUpperCase());
    }

    /**
     * Filtro de consulta: nulo ⇒ "sin filtrar", que es lo que necesita el
     * Calendario general (el modo que muestra Personal y Laboral juntos).
     * No confundir con {@link #fromNullable}, que resuelve el contexto de
     * ALTA de un recurso y ahí sí hay que decidirse por uno.
     */
    public static ModuleContext filterFromNullable(String raw) {
        if (raw == null || raw.isBlank()) {
            return null;
        }
        return ModuleContext.valueOf(raw.trim().toUpperCase());
    }
}

File: web/src/core/user/moduleContext.ts
/**
 * ADR-019 — aislamiento de recursos por módulo.
 *
 * Pedido explícito del usuario (2026-08-28): un recurso pertenece al módulo
 * desde el que se creó y no debe aparecer ni afectar al otro, "también en la
 * lógica de negocio y en las consultas/fuentes de datos", no solo en la
 * vista.
 *
 * Este archivo es solo el puente entre el modo activo del cliente y el
 * parámetro `context` de la API: el aislamiento de verdad está en el
 * servidor (columna `context` + consultas filtradas, migración V25). Sin el
 * servidor, esto no sería más que el filtro visual que el requisito prohíbe.
 */
export type ModuleContext = 'PERSONAL' | 'LABORAL'

/**
 * Añade `?context=` cuando hay módulo activo. Sin módulo (Calendario
 * general) se omite a propósito: ese modo existe para ver Personal y
 * Laboral juntos.
 */
export function withContext(path: string, context: ModuleContext | null | undefined): string {
  if (!context) return path
  return `${path}${path.includes('?') ? '&' : '?'}context=${context}`
}

/**
 * Contexto con el que se da de alta un recurso. Sin módulo activo el
 * recurso nace PERSONAL: es la regla 4 del pedido ("todos los recursos que
 * actualmente existen... pertenecen a Personal") aplicada también a las
 * pantallas que viven fuera de /personal y /laboral, como Garantías,
 * Mantenimiento o Suscripciones, que cuelgan del menú de Personal.
 */
export function creationContext(context: ModuleContext | null | undefined): ModuleContext {
  return context ?? 'PERSONAL'
}
